src/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs

```diff
@@ -116,10 +117,143 @@ private void Visit(IAssemblySymbol assembly)
+        private static bool IsOrContainsNonEmptyStruct(ITypeSymbol root)
+        {
+            HashSet<ITypeSymbol> visited = new(SymbolEqualityComparer.Default);
+            return WalkTypeSymbol(root, visited, ty => 
+                ty.IsUnmanagedType || 
+                    ((ty.IsReferenceType || ty.IsRefLikeType) && !SymbolEqualityComparer.Default.Equals(root, ty)));
+        }
+
+        // Convert IEnumerable<AttributeData> to a SyntaxList<AttributeListSyntax>.
+        private static SyntaxList<AttributeListSyntax> FromAttributeData(IEnumerable<AttributeData> attrData)
+        {
+            IEnumerable<SyntaxNode?> syntaxNodes = attrData.Select(ad =>
+                ad.ApplicationSyntaxReference?.GetSyntax(new System.Threading.CancellationToken(false)));
+            IEnumerable<AttributeListSyntax?> asNodes = syntaxNodes.Select(sn => {
+                if (sn is AttributeSyntax ass) {
+                    SeparatedSyntaxList<AttributeSyntax> singletonList = SyntaxFactory.SingletonSeparatedList<AttributeSyntax>(ass);
+                    AttributeListSyntax alSyntax = SyntaxFactory.AttributeList(singletonList);
+                    return alSyntax;
+                }
+
+                return null;
+            });
+            List<AttributeListSyntax> asList = asNodes.Where(a => a != null).OfType<AttributeListSyntax>().ToList();
+            return SyntaxFactory.List(asList);
+        }
+
+        // Build dummy field from a type, field name, and attribute list.
+        private static SyntaxNode CreateDummyField(string typ, string fieldName, SyntaxList<AttributeListSyntax> attrs)
+        {
+            return SyntaxFactory.FieldDeclaration(
+                SyntaxFactory.VariableDeclaration(
+                    SyntaxFactory.ParseTypeName(typ))
+                .WithVariables(
+                    SyntaxFactory.SingletonSeparatedList<Microsoft.CodeAnalysis.CSharp.Syntax.VariableDeclaratorSyntax>(
+                        SyntaxFactory.VariableDeclarator(
+                            SyntaxFactory.Identifier(fieldName))))
+            ).WithModifiers(SyntaxFactory.TokenList(new []{SyntaxFactory.Token(SyntaxKind.PrivateKeyword)})
+            ).WithAttributeLists(attrs);
+        }
+
+        // SynthesizeDummyFields yields private fields for the namedType, because they can be part of the API contract.
+        // - A struct containing a field that is a reference type cannot be used as a reference.
+        // - A struct containing nonempty fields needs to be fully initialized. (See "definite assignment" rules)
+        //   - "non-empty" means either unmanaged types like ints and enums, or reference types that are not the root.
+        // - A struct containing generic fields cannot have struct layout cycles.
+        private static IEnumerable<SyntaxNode> SynthesizeDummyFields(ISymbolFilter symbolFilter, INamedTypeSymbol namedType)
```

Comment:
Suggest moving the logic into separate source file. We have the `INamedTypeSymbolExtension` class and looks as a good place for this method (as an extension method)